File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) % 2 != 0:
            return False
        arr = []
        open = ['(', '{', '[']
        for el in s:
            if el in open:
                arr.append(el)
            else:
                if len(arr) == 0:
                    return False
                if el == ')':
                    if arr[-1] == '(':
                        del arr[-1]
                    else:
                        return False
                elif el == '}':
                    if arr[-1] == '{':
                        del arr[-1]
                    else:
                        return False
                elif el == ']':
                    if arr[-1] == '[':
                        del arr[-1]
                    else:
                        return False
        
        return (len(arr) == 0)
                     
        
                

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String str) {
        HashMap<Character, Character> close = new HashMap<>();
        close.put(')', '(');
        close.put('}', '{');
        close.put(']', '[');
        
        Set<Character> open = new HashSet<>();
        open.add('(');
        open.add('{');
        open.add('[');
        
        Stack<Character> stack = new Stack<>();
        char ch;

        for (int i = 0; i < str.length(); i++) {
            ch = str.charAt(i);
            if (open.contains(ch)) {
                stack.add(ch);
            } else if (close.containsKey(ch)) {
                if (stack.isEmpty()) {
                    return false;
                } else if (close.get(ch) != stack.pop()) {
                    return false;
                }
            }
        }
        return stack.isEmpty();
    }
}